### PR TITLE
Redirect & Page Not Found

### PR DESCRIPTION
**Pull request by fi-ya**
This PR focuses on `Simple redirect` and 'Page not found` features of acceptance tests
 **Scenario**: Getting a resource that has been moved to a different location
    !  Given I make an GET request to "/redirect"                                                        
    ~  Then my response should have status code 301
    ~  And my response should have a location header with the "http://0.0.0.0:5000/simple_get" URI
    ~  And my response should have an empty body

 **Scenario**: Getting an unconfigured resource returns a 404
    !  Given I make a GET request to a page that does not exist     
    ~  Then my response should have status code 404
    
Changes have been made to `Router`, `RedirectHandler`, `PageNotFoundHandler` and their test classes.

### CODE REVIEW

File: src/main/java/org/httpserver/handler/RedirectHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    private String handleStatusLine(Request request) {
+        String httpVersion = request.getHttpVersion();
+        String statusCode = StatusCode.MOVED_PERMANENTLY.getStatusCode();
+        String statusText = String.valueOf(StatusCode.MOVED_PERMANENTLY).replace("_", " ");
+
+        return String.format("%s %s %s", httpVersion, statusCode, statusText);
+    }
+
+    private String handleHeaders(Request request) {
+        return "Location: http://127.0.0.1:5000/simple_get";
```
Comment (andy8thlight) [inline src/main/java/org/httpserver/handler/RedirectHandler.java:37]: Is there a way to make this line a bit more generic? It seems very specific.

File: src/test/java/org/httpserver/request/RequestParserTest.java [hunk 2]
```diff
@@ -28,8 +28,10 @@ void parseRequestThatHasRequestLineOnly() throws IOException {
 
     @Test
     void parseRequestThatHasRequestLineAndHeadersOnly() throws IOException {
-        String mockRequestString = "GET /simple_get HTTP/1.1\n" +
-                "Host: 0.0.0.0:5000\n";
+        String mockRequestString = """
+                GET /simple_get HTTP/1.1
+                Host: 0.0.0.0:5000
+                """;
```
Comment (andy8thlight) [inline src/test/java/org/httpserver/request/RequestParserTest.java:34]: OOh, nice, a multi-line String in Java. Shiny, never used this in Java before. Cool!

File: src/test/java/org/httpserver/server/RouterTest.java [hunk 3]
```diff
@@ -106,4 +106,30 @@ void returnPostHandler_whenEchoBodyRoute() {
...
         assertEquals(EchoBodyHandler.class, actualHandler.getClass());
     }
+
+    @Test
+    void returnRedirectHandler_whenRedirectRoute() {
+        requestLineStub.put("httpMethod", "GET");
+        requestLineStub.put("requestTarget", "/redirect");
+        Request requestMock = new Request(requestLineStub, requestHeadersStub, requestBodyStub);
+        Router router = new Router();
+
```
Comment (andy8thlight) [inline src/test/java/org/httpserver/server/RouterTest.java:116]: Some test duplication here that you could refactor out. This pattern maybe in other places and you could potentially reduce the boilerplate a bit.

**Review by andy8thlight — CHANGES_REQUESTED** [review]
This is nice work generally although we've got some hard-coding here which I think needs resolving.

I'm noticing that duplication is starting to build up which is some technical debt. It can be resolved fairly easily and make the code less error prone. Hint: the handlers.

File: src/main/java/org/httpserver/handler/RedirectHandler.java (hunk 1 above)
Comment (fi-ya) [reply src/main/java/org/httpserver/handler/RedirectHandler.java:37]: Do you mean something like this?

File: src/test/java/org/httpserver/server/RouterTest.java (hunk 3 above)
Comment (fi-ya) [reply src/test/java/org/httpserver/server/RouterTest.java:116]: i extracted out the duplicated code like so:
```
private Handler getActualHandler() {
        Request requestMock = new Request(requestLineStub, requestHeadersStub, requestBodyStub);
        Router router = new Router();

        return router.getHandler(requestMock);
    }
 ```
 
 So called `getActualHandler` in the test like so:
 ```
     @Test
    void returnRedirectHandler_whenRedirectRoute() {
        requestLineStub.put("httpMethod", "GET");
        requestLineStub.put("requestTarget", "/redirect");

        Handler actualHandler = getActualHandler();

        assertEquals(RedirectHandler.class, actualHandler.getClass());
    }
```

**Pull request merged**